src/DomainHost.sol

```diff
@@ -0,0 +1,490 @@
+    /// @param wad The amount of DAI to release [WAD]
+    function release(uint256 wad) external auth {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+
+        // Fix any permissionless repays that may have occurred
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(this));
+        if (art < ink) {
+            address _vow = vow;
+            uint256 diff;
+            unchecked {
+                diff = ink - art;
+            }
+            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+            vat.grab(ilk, address(this), address(this), _vow, 0, _int256(diff));
+        }
+
+        require(dai.transferFrom(escrow, address(this), wad), "DomainHost/transfer-failed");
+        daiJoin.join(address(this), wad);
+        int256 repay = -_int256(wad);
+        vat.frob(ilk, address(this), address(this), address(this), repay, repay);
+        vat.slip(ilk, address(this), repay);
+
+        grain -= wad;
+
+        emit Release(wad);
+    }
+
+    /// @notice Guest is pushing a surplus
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to account as ddai [WAD]
+    function _surplus(uint256 _lid, uint256 wad) internal ordered(_lid) {
+        ddai += wad;
+
+        emit Surplus(wad);
+    }
+
+    /// @notice Guest is pushing a deficit
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to account as dsin [WAD]
+    function _deficit(uint256 _lid, uint256 wad) internal ordered(_lid) {
+        dsin += wad;
+
+        emit Deficit(wad);
+    }
+
+    /// @notice Governance is accruing surplus
+    /// @dev This is a potentially dangerous operation as a malicious domain can trigger unintended consequences due to fake surplus
+    /// @param _grain It should be the amount of grain necessary to cover debt in the guest domain [WAD]
+    function accrue(uint256 _grain) external auth returns (uint256 _wad) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+        _wad = ddai;
+        require(_wad > 0, "DomainHost/no-surplus");
+
+        int256 diff = _int256(_grain) - _int256(grain);
+        if (diff > 0) {
+            vat.slip(ilk, address(this), diff);
+            vat.frob(ilk, address(this), address(this), address(this), diff, diff);
+            daiJoin.exit(escrow, uint256(diff));
+            grain = _grain;
+        }
+
+        dai.transferFrom(address(escrow), address(this), _wad);
+        daiJoin.join(address(vow), _wad);
+
+        ddai = 0;
+
+        emit Accrue(_wad);
+    }
+
+    /// @notice Move bad debt from the remote domain into the local vow
+    /// @dev This is a potentially dangerous operation as a malicious domain can drain the entire surplus buffer
+    /// Because of this we require an authed party to perform this operation
+    function _rectify() internal auth returns (uint256 _rid, uint256 _wad) {
```

Comment:
Since `rectify` is considered dangerous and it is authed with the intention that calling it should require the governance to manually asses the validity of 'help' requested by the guest domain, is it really that different than allowing governance to rectify arbitrary amount "blindly", irrespective of what is the actual bad debt of the domain?

That would allow us to get rid of dsin variables and a few methods on both sides without making the overall process of managing debt any more complex (manuall assesment is necessary anyway).

Reply:
I guess it reduces the complexity for normal operations to have the functions. So excepting something goes wrong it is less code that the spell will need to output. However I can tell you that I just found some vulnerability in `rectify` as in `accrue`. The governance action might be front run from a malicious domain, basically adding up more ddai or dsin than the one governance was expecting when creating the spell. Then it would depend on what happens between the governance delay against the time the tx takes to be executed in L1.
So probably that is a point in your favor to reduce complexity and just add an extra param to fix it completely with a custom number.
cc @hexonaut @olivdb

Reply:
Yeah as we discussed in the other PR there should be a `maxAmount` parameter to protect against frontrun as well as give governance more control.

Reply:
This has been added https://github.com/makerdao/dss-bridge/pull/1/commits/ed5d015933c071927730a8a60f811310b1147472